test/jdk/java/awt/ScrollPane/ScrollPaneRemoveAdd.java

```diff
@@ -102,14 +103,66 @@ public void start() throws Exception {
             robot.delay(50);
             robot.waitForIdle();
             if (!actionSema.getState()) {
-                throw new RuntimeException("ScrollPane doesn't handle correctly add after remove");
+                throw new RuntimeException("ScrollPane doesn't handle " +
+                        "correctly add after remove");
             }
         } finally {
-            EventQueue.invokeLater(() -> {
+            EventQueue.invokeAndWait(() -> {
                 if (frame != null) {
                     frame.dispose();
                 }
             });
         }
     }
+
+    private static class Semaphore {
+        volatile boolean state = false;
+        Object lock = new Object();
+        volatile int waiting = 0;
```

Comment:
Why are both `state` and `waiting` marked as volatile? Either is accessed from synchronized blocks protected by `lock` except for `getState` which should also use the `lock` object.

Reply:
I don't think its required to be volatile, not sure.

Reply:
They shouldn't be volatile, a proper synchronisation is used. But `getState` should use `lock` lock to return the state because access to `state` is synchronised using `lock`, not `this`.